storage/grpc_client.go

```diff
@@ -1492,10 +1551,144 @@ func (r *gRPCReader) recv() (*storagepb.ReadObjectResponse, error) {
 		// reopen the stream, but will backoff if further attempts are necessary.
 		// Reopening the stream Recvs the first message, so if retrying is
 		// successful, the next logical chunk will be returned.
-		msg, err = r.reopenStream()
+		msg, err := r.reopenStream()
+		return msg.GetChecksummedData().GetContent(), err
+	}
+
+	if err != nil {
+		return nil, err
+	}
+
+	return readObjectResponseContent(r.databuf)
+}
+
+// readObjectResponseContent returns the checksummed_data.content field of a
+// ReadObjectResponse message, or an error if the message is invalid.
+// This can be used on recvs of objects after the first recv, since only the
+// first message will contain non-data fields.
+func readObjectResponseContent(b []byte) ([]byte, error) {
+	const (
+		readObjectResponse_checksummedData = protowire.Number(1)
+		checksummedData_content            = protowire.Number(1)
+	)
+
+	checksummedData, err := readProtoBytes(b, readObjectResponse_checksummedData)
+	if err != nil {
+		return b, fmt.Errorf("invalid ReadObjectResponse: %v", err)
+	}
+	content, err := readProtoBytes(checksummedData, checksummedData_content)
+	if err != nil {
+		return content, fmt.Errorf("invalid ReadObjectResponse: %v", err)
+	}
+
+	return content, nil
+}
+
+// readFullObjectResponse returns the ReadObjectResponse encoded in the
+// wire-encoded message buffer b, or an error if the message is invalid.
+// This is used on the first recv of an object as it may contain all fields of
+// ReadObjectResponse.
+func readFullObjectResponse(b []byte) (*storagepb.ReadObjectResponse, error) {
+	const (
+		checksummedDataField        = protowire.Number(1)
+		checksummedDataContentField = protowire.Number(1)
+		checksummedDataCRC32CField  = protowire.Number(2)
+		objectChecksumsField        = protowire.Number(2)
+		contentRangeField           = protowire.Number(3)
+		metadataField               = protowire.Number(4)
+	)
+
+	checksummedData, err := readProtoBytes(b, checksummedDataField)
+	if err != nil {
+		return nil, fmt.Errorf("invalid ReadObjectResponse: %v", err)
```

Comment:
These errors should include the field name in the failure; otherwise it's hard to diagnose what's going on.